extra_data/tests/test_keydata.py

```diff
@@ -284,6 +285,22 @@ def test_single_value(mock_sa3_control_data, monkeypatch):
     np.testing.assert_equal(intensity.as_single_value(rtol=1), np.median(data))
 
 
+def test_ndarray_out(mock_sa3_control_data):
+    f = H5File(mock_sa3_control_data)
+    beamview = f['SA3_XTD10_IMGFEL/CAM/BEAMVIEW2:daqOutput', 'data.image.dims']
+
+    # Get data and get a new array.
+    buf_new = beamview.ndarray()
+
+    # Get data into a pre-allocated array.
+    buf_in = np.empty(  # Do not use .zeros here to get random data!
+        beamview.shape[:1] + beamview.entry_shape, dtype=beamview.dtype)
+    buf_out = beamview.ndarray(out=buf_in)
+
+    np.testing.assert_allclose(buf_new, buf_out)
+    assert np.may_share_memory(buf_in, buf_out)
```

Comment:
TIL about this function. :slightly_smiling_face: 

I think we could just `assert buf_out is buf_in`, though?

Reply:
That is an interesting question, is that part of its API? I've always considered the `out=` argument to indicate memory to use, not necessarily exactly that object. But you're probably right.

Reply:
Interestingly the `numpy` documentation is not 100% explicit here. Taking [`numpy.add`](https://numpy.org/doc/stable/reference/generated/numpy.add.html) as an example:

> out: ndarray, None, or tuple of ndarray and None, optional
> A location into which the result is stored. [...] If not provided or None, a freshly-allocated array is returned.

> Returns: add: ndarray or scalar
> The sum of x1 and x2, element-wise.

The choice of words does not hint that the returned array at any point is the same one as passed to `out=`, only that there is no additional array allocation (is a new view an array allocation?). It leaves room for the returned value to possibly be of different type, or have additional annotations without breaking the API.

Reply:
Interesting - I've always taken it to mean that the same array object is returned. There's no obvious need to create a new view if you're not changing the shape or the dtype. :shrug:

Reply:
So would you like to make that part of the interface? I am fine either way, so far I was just under the assumption this is not guaranteed.

Reply:
I think I'm happy enough with that to make the test more specific - I can't imagine any reason we'd need to return a different view of the array that's passed in.